src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/RazorSemanticTokensEndpoint.cs

```diff
@@ -7,45 +7,32 @@
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
 using Microsoft.AspNetCore.Razor.LanguageServer.Semantic.Models;
-using Microsoft.Extensions.Logging;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic
 {
     internal class RazorSemanticTokensEndpoint : ISemanticTokensRangeEndpoint
     {
-        private readonly ILogger _logger;
-        private readonly RazorSemanticTokensInfoService _semanticTokensInfoService;
+        public bool MutatesSolutionState => false;
 
-        public RazorSemanticTokensEndpoint(
-            RazorSemanticTokensInfoService semanticTokensInfoService,
-            ILoggerFactory loggerFactory)
+        public RazorSemanticTokensEndpoint()
         {
-            if (semanticTokensInfoService is null)
-            {
-                throw new ArgumentNullException(nameof(semanticTokensInfoService));
-            }
-
-            if (loggerFactory is null)
-            {
-                throw new ArgumentNullException(nameof(loggerFactory));
-            }
-
-            _semanticTokensInfoService = semanticTokensInfoService;
-            _logger = loggerFactory.CreateLogger<RazorSemanticTokensEndpoint>();
         }
 
-        public async Task<SemanticTokens?> Handle(SemanticTokensRangeParamsBridge request, CancellationToken cancellationToken)
+        public async Task<SemanticTokens?> HandleRequestAsync(SemanticTokensRangeParamsBridge request, RazorRequestContext context, CancellationToken cancellationToken)
         {
             if (request is null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
 
-            var semanticTokens = await _semanticTokensInfoService.GetSemanticTokensAsync(request.TextDocument, request.Range, cancellationToken);
+            var documentContext = context.GetRequiredDocumentContext();
+            var semanticTokensInfoService = context.GetRequiredService<RazorSemanticTokensInfoService>();
+
+            var semanticTokens = await semanticTokensInfoService.GetSemanticTokensAsync(request.TextDocument, request.Range, documentContext, cancellationToken);
             var amount = semanticTokens is null ? "no" : (semanticTokens.Data.Length / 5).ToString(Thread.CurrentThread.CurrentCulture);
 
-            _logger.LogInformation("Returned {amount} semantic tokens for range {request.Range} in {request.TextDocument.Uri}.", amount, request.Range, request.TextDocument.Uri);
+            context.LspLogger.LogInformation($"Returned {amount} semantic tokens for range {request.Range} in {request.TextDocument.Uri}.");
```

Comment:
```suggestion
            context.Logger.LogInformation($"Returned {amount} semantic tokens for range {request.Range} in {request.TextDocument.Uri}.");
```

I'm not sure how you're doing this PR, but knowing you have a branch that compiles, I feel like there must be a better way. I would have hoped that updating a branch so it is on latest main, and compiles, and then staging specific changes progressively would work.. but maybe not? 🤷‍♂️